### ISSUE TITLE
How to add "unknown" class?

### ISSUE THREAD

**Issue by avikj** [davidsandberg/facenet#317]
I'm currently classifying aligned faces as one of 3 different people. Many of the images that the model will be supplied will not be any of those 3. I considered adding a class with random assorted training data, but I don't think that makes sense with an SVM.

How would I go about doing this?

**Comment by zo7** [comment]
FaceNet learns the embeddings of faces such that two embeddings of the same face have a small l2 distance (and different faces are far away to each other), reducing face recognition/verification to a simple nearest neighbors problem. For what you're doing, you can simply threshold the distance of a test image from several reference images that you have of the person you want to check for to see if it's that person or not, you don't need to train any models.

**Comment by zixia** [comment]
@avikj How about your progress, I have the same idea as yours. +1

**Comment by pavankumarkatakam** [comment]
I'm also trying to implement this.

**Comment by avikj** [comment]
Thanks @zo7, thats what I ended up doing and it works pretty well. @zixia @pavankumarkatakam Try using L2 distance to threshold; another thing to keep in mind is facenet can't produce good embeddings for very low quality images, where even a human would have trouble making out the face. Using image size, blur detection etc can also help filter out the unknowns.

**Comment by pavankumarkatakam** [comment]
@avikj, @zo7 I know that L2 distance can be used to classify and with respect to dlib the embedding distance threshold is 0.6 means if tow embeddings distance is less than 0.6 then they are same person face or else not. Can anybody suggest me the threshold value for facenet embeddings?

But if we have so many faces to recognise (100000) identities simply looping the every identity and checking the distance is not a good idea so I'm using the sci-kit-learn [KNN](http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html#sklearn.neighbors.KNeighborsClassifier) algorithm to search the nearest distance spatial embeddings and its face (Please correct me if I'm wrong in this decision) with parameters `metric="minkowski"` and `p=2` which is indirectly it is checking with respect to L2 distance.

please check this [issue](https://github.com/davidsandberg/facenet/issues/375) that my results on my own dataset. But the problem is for unknown person also sometimes im getting the 1.0 confidence. That is may be the problem with low-quality images and video frames im using because im using the Laptop inbuilt camera module which has only 2mp I think. I will try with 32mp camera with high resolution.

Finally two questions:
  1. is KNN can be used for L2 distance if more than 5000 identities to recognize.
  2. Is there any difference between DLIB embedding and facenet embedding.

**Comment by avikj** [comment]
@pavankumarkatakam you can still just check the distance to determine unknowns, even when using KNN. Once you identify the most likely class, you can find the average difference between a test point and that class; if it's above a threshold, you can say that it is unknown. So instead of checking distance against all classes, just check against the most likely class. KNN queries can be slow with large datasets. There are approximate nearest neighbor algorithms which work reasonably well and much faster than KNN; see LSH Forests if you are using sklearn.

**Comment by pavankumarkatakam** [comment]
Thanks for the info @avikj, first thing is to check the LSH Forests (which is faster than the KNN) and use distance calculation on likely classes rather than calculating on all classes. 
Also let me if you have any ideas on DLIB embeddings.

**Comment by avikj** [comment]
When setting a threshold value, there is a tradeoff between precision and recall. A lower threshold means there will be less false positives, but more false negatives, with the opposite being true for a high threshold. I used a threshold of ~.85 for my use case, where precision was more important than recall. I haven't looked into DLIB.

**Comment by pavankumarkatakam** [comment]
Thanks may I know which camera you are using

**Comment by avikj** [comment]
Not using a camera, my use case is on web videos.

**Comment by pavankumarkatakam** [comment]
ok Thanks

**Comment by niu-niu** [comment]
how about k-d tree search? @avikj

**Comment by varshneyabhi** [comment]
@avikj How are you deciding Anchor embedding to calculate distance? I tried the same trick. I searched for Embedding from which all distance are low (usually median) and used it to calculate distance with images needs to be predicted. Still distance for some Unknown people is less than 0.60. Am I doing something wrong? Please suggest.

**Comment by chungyau97** [comment]
Hi I'm a noob, I really have no clue what you guys are talking about... 
can anyone guide me? I really want to learn it. :(
